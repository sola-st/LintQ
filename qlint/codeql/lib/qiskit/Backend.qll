import python
import semmle.python.dataflow.new.DataFlow
import semmle.python.ApiGraphs
import qiskit.Register
import qiskit.Gate
import qiskit.Circuit

/** Simulator backend. */
class Backend extends DataFlow::CallCfgNode {
  // case: from qiskit import Aer; Aer.get_backend('qasm_simulator')
  Backend() {
    this = API::moduleImport("qiskit").getMember("Aer").getMember("get_backend").getACall() or
    this = API::moduleImport("qiskit").getMember("BasicAer").getMember("get_backend").getACall()
  }

  /** The name of the backend (according to the initialization string). */
  string getKind() {
    // case: from qiskit import Aer; Aer.get_backend('qasm_simulator')
    // > 'qasm_simulator'
    exists(StrConst str, DataFlow::LocalSourceNode strSource |
      strSource = this.getArg(0) and
      strSource.asExpr() = str
    |
      result = str.getText()
    )
  }

  /** If the simulator is statevector simulator. */
  predicate isStatevectorSimulator() {
    // case: from qiskit import Aer; Aer.get_backend('statevector_simulator')
    this.getKind().matches("%statevector%")
  }

  /** If the simulator is Unitary simulator. */
  predicate isUnitarySimulator() {
    // case: from qiskit import Aer; Aer.get_backend('unitary_simulator')
    this.getKind().matches("%unitary%")
  }

  /** Get a circuit run with this backend. */
  QuantumCircuit getACircuitToBeRun() {
    // case: Aer.get_backend('qasm_simulator').run(circuit)
    // > circuit
    exists(QuantumCircuit circuit, DataFlow::CallCfgNode runCall |
      runCall = this.getAnAttributeRead("run").getACall() and
      (
        circuit.flowsTo(runCall.getArg(0)) or
        circuit.flowsTo(runCall.getArgByName("circuits"))
      )
    |
      result = circuit
    )
    or
    // case: execute(qc, Aer.get_backend('qasm_simulator'), shots=1)
    // > qc
    exists(QuantumCircuit circuit, ExecuteCall executeCall |
      (
        circuit.flowsTo(executeCall.getArg(0)) or
        circuit.flowsTo(executeCall.getArgByName("experiments"))
      ) and
      (
        this.flowsTo(executeCall.getArg(1)) or
        this.flowsTo(executeCall.getArgByName("backend"))
      )
    |
      result = circuit
    )
  }

  /** Get a backend run with this backend. */
  BackendRun getABackendRun() {
    exists(BackendRun bkdRun |
      bkdRun.getBackend() = this
    |
      result = bkdRun
    )
  }

}

/** Backend run.
 *
 * This is a call to a backend run method.
 */
abstract class BackendRun extends DataFlow::CallCfgNode {

  abstract Backend getBackend();

}

/** A call to a backend run method. */
class BackendRunViaRunCall extends BackendRun {
  BackendRunViaRunCall() {
    exists(Backend backend, DataFlow::CallCfgNode runCall |
      runCall = backend.getAnAttributeRead("run").getACall()
    |
      this = runCall
    )
  }

  /** The backend that is run. */
  override Backend getBackend() {
    exists(Backend backend, DataFlow::CallCfgNode runCall |
      runCall = backend.getAnAttributeRead("run").getACall()
    |
      result = backend
    )
  }
}

/** A call to a backend execute method. */
class BackendRunViaExecuteCall extends BackendRun {
  BackendRunViaExecuteCall() {
    this instanceof ExecuteCall
  }

  /** The backend that is run. */
  override Backend getBackend() {
    exists(Backend backend |
      (
        backend.flowsTo(this.getArg(1)) or
        backend.flowsTo(this.getArgByName("backend"))
      )
    |
      result = backend
    )
  }
}

/** Result of a simulator run. */
class BackendResult extends DataFlow::CallCfgNode {
  BackendResult() {
    exists(BackendRun bkdRun |
      this = bkdRun.getAnAttributeRead("result").getACall()
      or
      this = bkdRun.getAnAttributeReference("result")
      or
      this = bkdRun.getAnAttributeWrite("result")
    )
  }

  /** The backend run that generated this result. */
  BackendRun getBackendRun() {
    exists(BackendRun bkdRun |
      // connect the run and its result
      this = bkdRun.getAnAttributeRead("result").getACall() or
      this = bkdRun.getAnAttributeReference("result") or
      this = bkdRun.getAnAttributeWrite("result")
    |
      result = bkdRun
    )
  }

}


/** Statevector returned by a result of a simulator run. */
class Statevector extends DataFlow::CallCfgNode {
  Statevector() {
    exists(BackendResult bkdResult |
      this = bkdResult.getAnAttributeRead("get_statevector").getACall()
    )
  }

  /** The backend run that generated this statevector. */
  BackendRun producedByBackendRun() {
    exists(BackendResult bkdResult, BackendRun bkdRun |
      // connect the run and its result
      bkdResult.getBackendRun() = bkdRun and
      // the result has been called to extract the statevector
      this = bkdResult.getAnAttributeRead("get_statevector").getACall()
    |
      result = bkdRun
    )
  }

}