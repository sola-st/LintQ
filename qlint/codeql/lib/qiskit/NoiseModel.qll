
import python
import semmle.python.dataflow.new.DataFlow
import semmle.python.ApiGraphs


class PauliError extends DataFlow::CallCfgNode {

    PauliError() {
        // OLDER API
        // from qiskit.providers.aer.noise.errors import pauli_error
        this = API::moduleImport("qiskit").getMember("providers")
            .getMember("aer").getMember("noise").getMember("errors")
            .getMember("pauli_error").getACall()
        or
        // NEWER API
        this = API::moduleImport("qiskit_aer").getMember("noise")
            .getMember("pauli_error").getACall()
    }

    StrConst getPauliString() {
        // pauli_error([('XY', p), ('I', 1 - p)])
        // extract 'XY' and 'I'
        exists(
            StrConst strConst
            |
            this.getArg(0).asCfgNode().getNode().contains(strConst)
            |
            result = strConst
            )
    }

    /** Check that the Pauli String are of different size.*/
    predicate arePauliStringSameSize() {
        // pauli_error([('XY', p), ('I', 1 - p)])
        // extract 'XY' and 'I'
        not exists(
            string str1, string str2
            |
            this.getPauliString().getText() = str1
            and
            this.getPauliString().getText() = str2
            |
            str1.length() != str2.length()
            )
    }


}
