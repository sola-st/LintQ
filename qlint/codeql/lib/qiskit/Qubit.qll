
// import qiskit.Circuit
// import qiskit.Register
// import tutorial
// // QubitUsed
// // derived from a Quantum gate used and its argument
// // implicitely connected to a default QuantumRegister via an index
// // e.g. qc.h(0)
// // or explicitely connected to a QuantumRegister via quantum_register[0]
// // e.g. qc.h(quantum_register[0])
// // TODO rename QubitAccess
// class QubitUsedInteger extends IntegerLiteral {
//     QubitUsedInteger() {
//         exists(
//             Gate gate
//             |
//             this = gate.getATargetQubit()
//         )
//     }
//     Gate getGate() {
//         exists(
//             Gate gate
//             |
//             this = gate.getATargetQubit()
//             |
//             result = gate
//         )
//     }
//     int getQubitIndex() {
//         result = this.getValue()
//     }
//     QuantumRegister getQuantumRegister() {
//         // OLD
//         // exists(
//         //     QuantumCircuit circ, QuantumRegister qr
//         //     |
//         //     circ = this.getGate().getQuantumCircuit() and
//         //     qr = circ.getAQuantumRegister() and
//         //     count(QuantumRegister reg | reg = circ.getAQuantumRegister() | reg) = 1
//         //     |
//         //     result = qr
//         // )
//         // or
//         exists(
//             QuantumRegister qr, SubscriptNode subscr
//             |
//             qr.getVar().getAUse() = subscr.getObject()
//             and
//             this = subscr.getNode().getIndex()
//             |
//             result = qr
//         )
//         // NEW
//         // exists(
//         //     QuantumRegister qr,
//         //     QuantumCircuit circ,
//         //     Gate gate
//         //     |
//         //     // this qubit is used in a gate
//         //     this = gate.getATargetQubit()
//         //     // the register belongs to the circuit
//         //     and circ.getAQuantumRegister() = qr
//         //     // the gate belongs to the circuit
//         //     and circ = gate.getQuantumCircuit()
//         //     and (
//         //         // either it is the only register in the circuit
//         //         // count(circ.getAQuantumRegister()) = 1
//         //         // or
//         //         (
//         //             exists(
//         //                 SubscriptNode subscr
//         //                 |
//         //                 // or we look for the register variable which
//         //                 // is used to access the qubit (IntegerLiteral)
//         //                 qr.getVar().getAUse() = subscr.getObject()
//         //                 and this = subscr.getNode().getIndex()
//         //             )
//         //         )
//         //     )
//         //     |
//         //     result = qr
//         // )
//     }
// }
// QubitUsed
// derived from a Quantum gate used and its argument
// implicitely connected to a default QuantumRegister via an index
// e.g.  qc.h(0)
// or explicitely connected to a QuantumRegister via quantum_register[0]
// e.g. qc.h(quantum_register[0])
// .(Subscript).getIndex().(IntegerLiteral).getValue() = i
// class QubitUsed extends SubscriptNode {
//     // get a quantum register node which is used in a quantum gate
//     // qc.h(quantum_register[0])
//     QubitUsed() {
//         exists(
//             QuantumRegister qr, Variable reg_var
//             |
//             qr.getVar() = reg_var
//             and
//             reg_var.getAUse() = this.getObject()
//         ) or
//         exists(
//             QuantumRegister qr, Variable reg_var
//             |
//             qr.getVar() = reg_var
//             and
//             reg_var.getAUse() = this.getObject()
//         )
//     }
//     // get the index of the qubit used in the quantum gate
//     // qc.h(quantum_register[0])
//     int getQubitIndex() {
//         exists(
//             IntegerLiteral index
//             |
//             index = this.getNode().getIndex()
//             |
//             result = index.getValue()
//         )
//     }
//     QuantumRegister getQuantumRegister() {
//         exists(
//             QuantumRegister qr
//             |
//             qr.getVar().getAUse() = this.getObject()
//             |
//             result = qr
//         )
//     }
// }
// QubitSpace
// derived from a QuantumRegister declaration used in a QuantumCircuit
// e.g. quantum_register = QuantumRegister(2)
// Detector improvements
// find if there is any register with QubitSpace that is not used in any QubitUsed
