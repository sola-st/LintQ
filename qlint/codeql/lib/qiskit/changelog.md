2023.03.07
- fix missing gates `crx` and `cry` in the `Gate` class
- improved the `isAppliedAfter` in the `Gate` class to remove false positives when the circuit is re-initialized during the path connecting the two gates.

2023.03.07
- added the getVar() function to the register class to get the variable name of the register.
- added the QubitUsed abstraction to keep track of each access to a qubit, considering also the register it belongs to.

2023.03.03
- refactoring model: java style method calls
- refactoring of register to share part of it.
- removed GateQuantumOperation, substituted with a predicate isMeasurement() on a general gate.

2023.02.28
- refactoring of the Gate class. Remove the tracking of the for loop.

2023.02.21
- improve the efficiency of get_a_target_qubit() of the gate class (remove extra QuantumCircuit variable)
- improved the precision of get_a_target_qubit() of the gate class to account also for qubits which are operated on in a loop (e.g. qc.h(i)).


2023.02.20
- added is_subcircuit() method to Circuit class (by checking if it flows to the compose() or append() method of another circuit object)
- improved get_total_num_qubits() and get_total_num_bits() methods of Circuit class (by considering also the registers added add_register() method)
- in Circuit refactored get_num_qubits_with_integers() and get_num_bits_with_integers() methods to return the static size of the circuit (i.e. QuantumCircuit(2,2) or n = 7; m = 2, QuantumCircuit(n, m)).
- improved get_num_bits() and get_num_qubits() methods of Register class (by considering also local flows of integer variables, e.g. x = 5; QuantumRegister(x))