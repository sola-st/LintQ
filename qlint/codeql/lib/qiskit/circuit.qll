import python
import semmle.python.dataflow.new.DataFlow
import semmle.python.ApiGraphs
import qiskit.register


class QuantumCircuit extends DataFlow::CallCfgNode {
    QuantumCircuit() {
        this = API::moduleImport("qiskit").getMember("QuantumCircuit").getACall()
    }

    int get_num_qubits() {
        exists(IntegerLiteral num_qubits |
            num_qubits = this.getArg(0).asExpr() and
            result = num_qubits.getValue()
        ) or
        exists( QuantumRegister qntReg|
            // there is a dataflow between the quantum register and the quantum circuit
            exists(int i |
                qntReg.flowsTo(this.getArg(i)))
            and
            // the number of qubits is the number of qubits in the quantum register
            result = qntReg.get_num_qubits()
        )
    }

    int get_num_bits() {
        exists(IntegerLiteral num_bits |
            num_bits = this.getArg(1).asExpr() and
            result = num_bits.getValue()
        ) or
        exists( ClassicalRegister clsReg|
            // there is a dataflow between the classical register and the quantum circuit
            exists(int i |
                clsReg.flowsTo(this.getArg(i)))
            and
            // the number of bits is the number of bits in the classical register
            result = clsReg.get_num_bits()
        )
    }

    int get_total_num_bits() {
        result = sum(ClassicalRegister clsReg, int i |
            clsReg.flowsTo(this.getArg(i)) |
            clsReg.get_num_bits())
    }

    int get_total_num_qubits() {
        result = sum(QuantumRegister qntReg, int i |
            qntReg.flowsTo(this.getArg(i)) |
            qntReg.get_num_qubits())
    }

}