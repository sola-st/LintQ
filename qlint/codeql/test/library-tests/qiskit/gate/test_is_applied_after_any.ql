import python
import semmle.python.dataflow.new.DataFlow
import semmle.python.ApiGraphs
import qiskit.Circuit
import qiskit.Gate
import qiskit.Qubit

from Gate gateBefore, Gate gateAfter, int shared_qubit
where
  not gateBefore.getLocation().getFile().getAbsolutePath().matches("%site-packages/qiskit/%") and
  not gateAfter.getLocation().getFile().getAbsolutePath().matches("%site-packages/qiskit/%") and
  gateAfter.isAppliedAfterOn(gateBefore, shared_qubit)
select gateBefore, gateAfter,
  "Gate: '" + gateBefore.getGateName() + "' on qubit " + shared_qubit + " is followed by gate: '" +
    gateAfter.getGateName() + "'."
