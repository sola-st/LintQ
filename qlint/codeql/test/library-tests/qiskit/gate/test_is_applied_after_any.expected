| concept_reset_gate.py:8:1:8:13 | ControlFlowNode for Attribute() | concept_reset_gate.py:9:1:9:17 | ControlFlowNode for Attribute() | Gate: 'x' on qubit 0 is followed by gate: 'cx'. |
| concept_reset_gate.py:8:1:8:13 | ControlFlowNode for Attribute() | concept_reset_gate.py:10:1:10:22 | ControlFlowNode for Attribute() | Gate: 'x' on qubit 0 is followed by gate: 'measure'. |
| concept_reset_gate.py:8:1:8:13 | ControlFlowNode for Attribute() | concept_reset_gate.py:12:1:12:17 | ControlFlowNode for Attribute() | Gate: 'x' on qubit 0 is followed by gate: 'reset'. |
| concept_reset_gate.py:8:1:8:13 | ControlFlowNode for Attribute() | concept_reset_gate.py:14:1:14:13 | ControlFlowNode for Attribute() | Gate: 'x' on qubit 0 is followed by gate: 'h'. |
| concept_reset_gate.py:9:1:9:17 | ControlFlowNode for Attribute() | concept_reset_gate.py:10:1:10:22 | ControlFlowNode for Attribute() | Gate: 'cx' on qubit 0 is followed by gate: 'measure'. |
| concept_reset_gate.py:9:1:9:17 | ControlFlowNode for Attribute() | concept_reset_gate.py:11:1:11:22 | ControlFlowNode for Attribute() | Gate: 'cx' on qubit 1 is followed by gate: 'measure'. |
| concept_reset_gate.py:9:1:9:17 | ControlFlowNode for Attribute() | concept_reset_gate.py:12:1:12:17 | ControlFlowNode for Attribute() | Gate: 'cx' on qubit 0 is followed by gate: 'reset'. |
| concept_reset_gate.py:9:1:9:17 | ControlFlowNode for Attribute() | concept_reset_gate.py:13:1:13:17 | ControlFlowNode for Attribute() | Gate: 'cx' on qubit 1 is followed by gate: 'reset'. |
| concept_reset_gate.py:9:1:9:17 | ControlFlowNode for Attribute() | concept_reset_gate.py:14:1:14:13 | ControlFlowNode for Attribute() | Gate: 'cx' on qubit 0 is followed by gate: 'h'. |
| concept_reset_gate.py:10:1:10:22 | ControlFlowNode for Attribute() | concept_reset_gate.py:12:1:12:17 | ControlFlowNode for Attribute() | Gate: 'measure' on qubit 0 is followed by gate: 'reset'. |
| concept_reset_gate.py:10:1:10:22 | ControlFlowNode for Attribute() | concept_reset_gate.py:14:1:14:13 | ControlFlowNode for Attribute() | Gate: 'measure' on qubit 0 is followed by gate: 'h'. |
| concept_reset_gate.py:11:1:11:22 | ControlFlowNode for Attribute() | concept_reset_gate.py:13:1:13:17 | ControlFlowNode for Attribute() | Gate: 'measure' on qubit 1 is followed by gate: 'reset'. |
| concept_reset_gate.py:12:1:12:17 | ControlFlowNode for Attribute() | concept_reset_gate.py:14:1:14:13 | ControlFlowNode for Attribute() | Gate: 'reset' on qubit 0 is followed by gate: 'h'. |
| concept_reset_gate.py:19:1:19:17 | ControlFlowNode for Attribute() | concept_reset_gate.py:22:1:22:13 | ControlFlowNode for Attribute() | Gate: 'reset' on qubit 0 is followed by gate: 'x'. |
| concept_reset_gate.py:19:1:19:17 | ControlFlowNode for Attribute() | concept_reset_gate.py:23:1:23:17 | ControlFlowNode for Attribute() | Gate: 'reset' on qubit 0 is followed by gate: 'cx'. |
| concept_reset_gate.py:19:1:19:17 | ControlFlowNode for Attribute() | concept_reset_gate.py:24:1:24:22 | ControlFlowNode for Attribute() | Gate: 'reset' on qubit 0 is followed by gate: 'measure'. |
| concept_reset_gate.py:19:1:19:17 | ControlFlowNode for Attribute() | concept_reset_gate.py:26:1:26:13 | ControlFlowNode for Attribute() | Gate: 'reset' on qubit 0 is followed by gate: 'h'. |
| concept_reset_gate.py:20:1:20:17 | ControlFlowNode for Attribute() | concept_reset_gate.py:23:1:23:17 | ControlFlowNode for Attribute() | Gate: 'reset' on qubit 1 is followed by gate: 'cx'. |
| concept_reset_gate.py:20:1:20:17 | ControlFlowNode for Attribute() | concept_reset_gate.py:25:1:25:22 | ControlFlowNode for Attribute() | Gate: 'reset' on qubit 1 is followed by gate: 'measure'. |
| concept_reset_gate.py:22:1:22:13 | ControlFlowNode for Attribute() | concept_reset_gate.py:23:1:23:17 | ControlFlowNode for Attribute() | Gate: 'x' on qubit 0 is followed by gate: 'cx'. |
| concept_reset_gate.py:22:1:22:13 | ControlFlowNode for Attribute() | concept_reset_gate.py:24:1:24:22 | ControlFlowNode for Attribute() | Gate: 'x' on qubit 0 is followed by gate: 'measure'. |
| concept_reset_gate.py:22:1:22:13 | ControlFlowNode for Attribute() | concept_reset_gate.py:26:1:26:13 | ControlFlowNode for Attribute() | Gate: 'x' on qubit 0 is followed by gate: 'h'. |
| concept_reset_gate.py:23:1:23:17 | ControlFlowNode for Attribute() | concept_reset_gate.py:24:1:24:22 | ControlFlowNode for Attribute() | Gate: 'cx' on qubit 0 is followed by gate: 'measure'. |
| concept_reset_gate.py:23:1:23:17 | ControlFlowNode for Attribute() | concept_reset_gate.py:25:1:25:22 | ControlFlowNode for Attribute() | Gate: 'cx' on qubit 1 is followed by gate: 'measure'. |
| concept_reset_gate.py:23:1:23:17 | ControlFlowNode for Attribute() | concept_reset_gate.py:26:1:26:13 | ControlFlowNode for Attribute() | Gate: 'cx' on qubit 0 is followed by gate: 'h'. |
| concept_reset_gate.py:24:1:24:22 | ControlFlowNode for Attribute() | concept_reset_gate.py:26:1:26:13 | ControlFlowNode for Attribute() | Gate: 'measure' on qubit 0 is followed by gate: 'h'. |
| concept_reset_gate.py:34:5:34:15 | ControlFlowNode for Attribute() | concept_reset_gate.py:34:5:34:15 | ControlFlowNode for Attribute() | Gate: 'x' on qubit 2 is followed by gate: 'x'. |
| concept_reset_gate.py:34:5:34:15 | ControlFlowNode for Attribute() | concept_reset_gate.py:35:5:35:23 | ControlFlowNode for Attribute() | Gate: 'x' on qubit 2 is followed by gate: 'cx'. |
| concept_reset_gate.py:34:5:34:15 | ControlFlowNode for Attribute() | concept_reset_gate.py:38:5:38:16 | ControlFlowNode for Attribute() | Gate: 'x' on qubit 2 is followed by gate: 'reset'. |
| concept_reset_gate.py:35:5:35:23 | ControlFlowNode for Attribute() | concept_reset_gate.py:34:5:34:15 | ControlFlowNode for Attribute() | Gate: 'cx' on qubit 2 is followed by gate: 'x'. |
| concept_reset_gate.py:35:5:35:23 | ControlFlowNode for Attribute() | concept_reset_gate.py:35:5:35:23 | ControlFlowNode for Attribute() | Gate: 'cx' on qubit 1 is followed by gate: 'cx'. |
| concept_reset_gate.py:35:5:35:23 | ControlFlowNode for Attribute() | concept_reset_gate.py:35:5:35:23 | ControlFlowNode for Attribute() | Gate: 'cx' on qubit 2 is followed by gate: 'cx'. |
| concept_reset_gate.py:35:5:35:23 | ControlFlowNode for Attribute() | concept_reset_gate.py:37:5:37:28 | ControlFlowNode for Attribute() | Gate: 'cx' on qubit 1 is followed by gate: 'measure'. |
| concept_reset_gate.py:35:5:35:23 | ControlFlowNode for Attribute() | concept_reset_gate.py:38:5:38:16 | ControlFlowNode for Attribute() | Gate: 'cx' on qubit 1 is followed by gate: 'reset'. |
| concept_reset_gate.py:35:5:35:23 | ControlFlowNode for Attribute() | concept_reset_gate.py:38:5:38:16 | ControlFlowNode for Attribute() | Gate: 'cx' on qubit 2 is followed by gate: 'reset'. |
| concept_reset_gate.py:36:5:36:28 | ControlFlowNode for Attribute() | concept_reset_gate.py:36:5:36:28 | ControlFlowNode for Attribute() | Gate: 'measure' on qubit 0 is followed by gate: 'measure'. |
| concept_reset_gate.py:36:5:36:28 | ControlFlowNode for Attribute() | concept_reset_gate.py:38:5:38:16 | ControlFlowNode for Attribute() | Gate: 'measure' on qubit 0 is followed by gate: 'reset'. |
| concept_reset_gate.py:37:5:37:28 | ControlFlowNode for Attribute() | concept_reset_gate.py:35:5:35:23 | ControlFlowNode for Attribute() | Gate: 'measure' on qubit 1 is followed by gate: 'cx'. |
| concept_reset_gate.py:37:5:37:28 | ControlFlowNode for Attribute() | concept_reset_gate.py:37:5:37:28 | ControlFlowNode for Attribute() | Gate: 'measure' on qubit 1 is followed by gate: 'measure'. |
| concept_reset_gate.py:37:5:37:28 | ControlFlowNode for Attribute() | concept_reset_gate.py:38:5:38:16 | ControlFlowNode for Attribute() | Gate: 'measure' on qubit 1 is followed by gate: 'reset'. |
| concept_reset_gate.py:38:5:38:16 | ControlFlowNode for Attribute() | concept_reset_gate.py:34:5:34:15 | ControlFlowNode for Attribute() | Gate: 'reset' on qubit 2 is followed by gate: 'x'. |
| concept_reset_gate.py:38:5:38:16 | ControlFlowNode for Attribute() | concept_reset_gate.py:35:5:35:23 | ControlFlowNode for Attribute() | Gate: 'reset' on qubit 1 is followed by gate: 'cx'. |
| concept_reset_gate.py:38:5:38:16 | ControlFlowNode for Attribute() | concept_reset_gate.py:35:5:35:23 | ControlFlowNode for Attribute() | Gate: 'reset' on qubit 2 is followed by gate: 'cx'. |
| concept_reset_gate.py:38:5:38:16 | ControlFlowNode for Attribute() | concept_reset_gate.py:36:5:36:28 | ControlFlowNode for Attribute() | Gate: 'reset' on qubit 0 is followed by gate: 'measure'. |
| concept_reset_gate.py:38:5:38:16 | ControlFlowNode for Attribute() | concept_reset_gate.py:37:5:37:28 | ControlFlowNode for Attribute() | Gate: 'reset' on qubit 1 is followed by gate: 'measure'. |
| concept_reset_gate.py:38:5:38:16 | ControlFlowNode for Attribute() | concept_reset_gate.py:38:5:38:16 | ControlFlowNode for Attribute() | Gate: 'reset' on qubit 0 is followed by gate: 'reset'. |
| concept_reset_gate.py:38:5:38:16 | ControlFlowNode for Attribute() | concept_reset_gate.py:38:5:38:16 | ControlFlowNode for Attribute() | Gate: 'reset' on qubit 1 is followed by gate: 'reset'. |
| concept_reset_gate.py:38:5:38:16 | ControlFlowNode for Attribute() | concept_reset_gate.py:38:5:38:16 | ControlFlowNode for Attribute() | Gate: 'reset' on qubit 2 is followed by gate: 'reset'. |
| concept_reset_gate.py:48:5:48:22 | ControlFlowNode for Attribute() | concept_reset_gate.py:48:5:48:22 | ControlFlowNode for Attribute() | Gate: 'x' on qubit 2 is followed by gate: 'x'. |
| concept_reset_gate.py:48:5:48:22 | ControlFlowNode for Attribute() | concept_reset_gate.py:49:5:49:32 | ControlFlowNode for Attribute() | Gate: 'x' on qubit 2 is followed by gate: 'cx'. |
| concept_reset_gate.py:48:5:48:22 | ControlFlowNode for Attribute() | concept_reset_gate.py:53:5:53:23 | ControlFlowNode for Attribute() | Gate: 'x' on qubit 2 is followed by gate: 'reset'. |
| concept_reset_gate.py:49:5:49:32 | ControlFlowNode for Attribute() | concept_reset_gate.py:48:5:48:22 | ControlFlowNode for Attribute() | Gate: 'cx' on qubit 2 is followed by gate: 'x'. |
| concept_reset_gate.py:49:5:49:32 | ControlFlowNode for Attribute() | concept_reset_gate.py:49:5:49:32 | ControlFlowNode for Attribute() | Gate: 'cx' on qubit 1 is followed by gate: 'cx'. |
| concept_reset_gate.py:49:5:49:32 | ControlFlowNode for Attribute() | concept_reset_gate.py:49:5:49:32 | ControlFlowNode for Attribute() | Gate: 'cx' on qubit 2 is followed by gate: 'cx'. |
| concept_reset_gate.py:49:5:49:32 | ControlFlowNode for Attribute() | concept_reset_gate.py:51:5:51:37 | ControlFlowNode for Attribute() | Gate: 'cx' on qubit 1 is followed by gate: 'measure'. |
| concept_reset_gate.py:49:5:49:32 | ControlFlowNode for Attribute() | concept_reset_gate.py:53:5:53:23 | ControlFlowNode for Attribute() | Gate: 'cx' on qubit 1 is followed by gate: 'reset'. |
| concept_reset_gate.py:49:5:49:32 | ControlFlowNode for Attribute() | concept_reset_gate.py:53:5:53:23 | ControlFlowNode for Attribute() | Gate: 'cx' on qubit 2 is followed by gate: 'reset'. |
| concept_reset_gate.py:50:5:50:37 | ControlFlowNode for Attribute() | concept_reset_gate.py:50:5:50:37 | ControlFlowNode for Attribute() | Gate: 'measure' on qubit 0 is followed by gate: 'measure'. |
| concept_reset_gate.py:50:5:50:37 | ControlFlowNode for Attribute() | concept_reset_gate.py:53:5:53:23 | ControlFlowNode for Attribute() | Gate: 'measure' on qubit 0 is followed by gate: 'reset'. |
| concept_reset_gate.py:51:5:51:37 | ControlFlowNode for Attribute() | concept_reset_gate.py:49:5:49:32 | ControlFlowNode for Attribute() | Gate: 'measure' on qubit 1 is followed by gate: 'cx'. |
| concept_reset_gate.py:51:5:51:37 | ControlFlowNode for Attribute() | concept_reset_gate.py:51:5:51:37 | ControlFlowNode for Attribute() | Gate: 'measure' on qubit 1 is followed by gate: 'measure'. |
| concept_reset_gate.py:51:5:51:37 | ControlFlowNode for Attribute() | concept_reset_gate.py:53:5:53:23 | ControlFlowNode for Attribute() | Gate: 'measure' on qubit 1 is followed by gate: 'reset'. |
| concept_reset_gate.py:53:5:53:23 | ControlFlowNode for Attribute() | concept_reset_gate.py:48:5:48:22 | ControlFlowNode for Attribute() | Gate: 'reset' on qubit 2 is followed by gate: 'x'. |
| concept_reset_gate.py:53:5:53:23 | ControlFlowNode for Attribute() | concept_reset_gate.py:49:5:49:32 | ControlFlowNode for Attribute() | Gate: 'reset' on qubit 1 is followed by gate: 'cx'. |
| concept_reset_gate.py:53:5:53:23 | ControlFlowNode for Attribute() | concept_reset_gate.py:49:5:49:32 | ControlFlowNode for Attribute() | Gate: 'reset' on qubit 2 is followed by gate: 'cx'. |
| concept_reset_gate.py:53:5:53:23 | ControlFlowNode for Attribute() | concept_reset_gate.py:50:5:50:37 | ControlFlowNode for Attribute() | Gate: 'reset' on qubit 0 is followed by gate: 'measure'. |
| concept_reset_gate.py:53:5:53:23 | ControlFlowNode for Attribute() | concept_reset_gate.py:51:5:51:37 | ControlFlowNode for Attribute() | Gate: 'reset' on qubit 1 is followed by gate: 'measure'. |
| concept_reset_gate.py:53:5:53:23 | ControlFlowNode for Attribute() | concept_reset_gate.py:53:5:53:23 | ControlFlowNode for Attribute() | Gate: 'reset' on qubit 0 is followed by gate: 'reset'. |
| concept_reset_gate.py:53:5:53:23 | ControlFlowNode for Attribute() | concept_reset_gate.py:53:5:53:23 | ControlFlowNode for Attribute() | Gate: 'reset' on qubit 1 is followed by gate: 'reset'. |
| concept_reset_gate.py:53:5:53:23 | ControlFlowNode for Attribute() | concept_reset_gate.py:53:5:53:23 | ControlFlowNode for Attribute() | Gate: 'reset' on qubit 2 is followed by gate: 'reset'. |
| pred_is_applied_after.py:9:1:9:10 | ControlFlowNode for Attribute() | pred_is_applied_after.py:10:1:10:17 | ControlFlowNode for Attribute() | Gate: 'h' on qubit 0 is followed by gate: 'cx'. |
| pred_is_applied_after.py:15:1:15:13 | ControlFlowNode for Attribute() | pred_is_applied_after.py:16:1:16:17 | ControlFlowNode for Attribute() | Gate: 's' on qubit 0 is followed by gate: 'cx'. |
| pred_is_applied_after.py:16:1:16:17 | ControlFlowNode for Attribute() | pred_is_applied_after.py:17:1:17:17 | ControlFlowNode for Attribute() | Gate: 'cx' on qubit 1 is followed by gate: 'cx'. |
| pred_is_applied_after.py:17:1:17:17 | ControlFlowNode for Attribute() | pred_is_applied_after.py:18:1:18:17 | ControlFlowNode for Attribute() | Gate: 'cx' on qubit 2 is followed by gate: 'cx'. |
| pred_is_applied_after.py:18:1:18:17 | ControlFlowNode for Attribute() | pred_is_applied_after.py:19:1:19:13 | ControlFlowNode for Attribute() | Gate: 'cx' on qubit 3 is followed by gate: 't'. |
| pred_is_applied_after.py:43:1:43:16 | ControlFlowNode for Attribute() | pred_is_applied_after.py:44:1:44:18 | ControlFlowNode for Attribute() | Gate: 'measure' on qubit 2 is followed by gate: 'h'. |
